service/kernel/oneapi/cl_kernels/select_indexed.cl

```diff
@@ -46,86 +46,102 @@ void partition( __global  algorithmFPType *pdist,
-            int exLabel = flag > 0 ? plabels[start + *split + shift] : -1;
-            int cur_size = i > full_group_size - 1 ? last_group_size : local_size;
+            unsigned char shift     = sub_group_scan_exclusive_add(flag);
+            unsigned char old_shift = sub_group_scan_exclusive_add(flag > 0 ? 0 : 1);
+            int cur_size            = i > full_group_size - 1 ? last_group_size : local_size;
             if(flag) 
             {
-                pdist[start + *split + shift] = curVal;
+                algorithmFPType exVal           = pdist[start + *split + shift];
+                int exLabel                     = plabels[start + *split + shift];
+                pdist[start + *split + shift]   = curVal;
                 plabels[start + *split + shift] = curLabel;
-                if(start + *split + shift < min_ind) {
-                    pdist[min_ind + cur_size - 1 - shift] = exVal;
+                if(start + *split + shift < min_ind) 
+                {
+                    pdist[min_ind + cur_size - 1 - shift]   = exVal;
                     plabels[min_ind + cur_size - 1 - shift] = exLabel;
                 }
             } 
             else 
             {
-                pdist[min_ind + cur_size - 1 - old_shift] = curVal;
+                pdist[min_ind + cur_size - 1 - old_shift]   = curVal;
                 plabels[min_ind + cur_size - 1 - old_shift] = curLabel;
             }
-        }
+        } 
         sub_group_barrier(CLK_GLOBAL_MEM_FENCE);
-        *split += incr;
+        *split         += incr;
         *total_big_num += big_num;
     }
-    *split = - sub_group_reduce_min(-(*split));
+    *split         = - sub_group_reduce_min(-(*split));
     *total_big_num = - sub_group_reduce_min(-(*total_big_num));
 }
 
-__kernel void quick_select_group( __global  algorithmFPType           *distances,
-                            __global  int                       *labels,
-                            __global  algorithmFPType           *nbrdistances,
-                            __global  int                       *nbrlabels,
-                            __global  const algorithmFPType     *rndseq,
-                            int RndPeriod,
-                            int N,
-                            int K,
-                            int BlockOffset) 
+__kernel void quick_select_group(__global  algorithmFPType           *distances,
+                                 __global  int                       *labels,
+                                 __global  algorithmFPType           *nbrdistances,
+                                 __global  int                       *nbrlabels,
+                                 __global  const algorithmFPType     *rndseq,
+                                 int RndPeriod,
+                                 int N,
+                                 int NLast,
+                                 int K,
+                                 int BlockOffset) 
```

Comment:
If this is common quick select algorithm, why it is dependent of KNN-related inputs like `distances` or `labels`?